src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs

```diff
@@ -136,38 +142,25 @@ protected override async Task OnRun(CancellationToken token)
                         }
                     }
 
-                    DateTime currentTimestamp = _context.Clock.UtcNow.UtcDateTime;
+                    DateTime currentTimestamp = Context.Clock.UtcNow.UtcDateTime;
 
-                    // If rule has an action count window, Remove all execution timestamps that fall outside the window.
-                    if (actionCountWindowDuration.HasValue)
+                    if (_stateHolder.CanExecuteActions(currentTimestamp))
                     {
-                        DateTime windowStartTimestamp = currentTimestamp - actionCountWindowDuration.Value;
-                        while (executionTimestamps.Count > 0)
+                        lock (_stateHolder.ExecutionTimestamps)
                         {
-                            DateTime executionTimestamp = executionTimestamps.Peek();
-                            if (executionTimestamp < windowStartTimestamp)
-                            {
-                                executionTimestamps.Dequeue();
-                            }
-                            else
-                            {
-                                // Stop clearing out previous executions
-                                break;
-                            }
+                            _stateHolder.ExecutionTimestamps.Enqueue(currentTimestamp);
                         }
-                    }
 
-                    // Check if executing actions has been throttled due to count limit
-                    if (actionCountLimit > executionTimestamps.Count)
-                    {
-                        executionTimestamps.Enqueue(currentTimestamp);
+                        _stateHolder.AllExecutionTimestamps.Add(currentTimestamp);
```

Comment:
Since we clone AllexecutionTimestamps as well, we probably want to lock around this addition as well.